data/transactions/logic/eval.go

```diff
@@ -1096,14 +1091,56 @@ func EvalContract(program []byte, gi int, aid basics.AppIndex, params *EvalParam
 	// Save scratch for `gload`. We used to copy, but cx.scratch is quite large,
 	// about 8k, and caused measurable CPU and memory demands.  Of course, these
 	// should never be changed by later transactions.
-	cx.pastScratch[cx.groupIndex] = &cx.Scratch
+	cx.pastScratch[cx.groupIndex] = cx.Scratch
 
-	return pass, &cx, err
+	return pass, cx, err
+}
+
+var cxPool = sync.Pool{
+	New: func() interface{} {
+		return &EvalContext{}
+	},
+}
+
+func getEvalContext(gi int, aid basics.AppIndex, params *EvalParams) *EvalContext {
```

Comment:
This function is what bugs me about using `sync.Pool` for complicated structures.  If you add a field, you have to remember to modify this.

I worry about lines like `cx.Stack = cx.Stack[:0]`. This doesn't clear the values above 0, so they can not be GC'd. So if you have a lot of big  `[]byte` stack values, they stick around.

Line 1123 is the new allocation I was worried about.

Reply:
I agree, it logically belongs to txgroup and  EvalParams. Needs more thought/work, sure.